core_v14/src/decoder/mod.rs

```diff
@@ -254,50 +268,148 @@ impl Decoder {
 		log::trace!("pallet index: {}, call index: {}", pallet_index, call_index);
 
 		// Work out which call the extrinsic data represents and get type info for it:
-		let (pallet_name, call) = match self.metadata.call_by_variant_index(pallet_index, call_index) {
+		let (pallet_name, variant) = match self.metadata.call_variant_by_enum_index(pallet_index, call_index) {
 			Some(call) => call,
 			None => return Err(DecodeError::CannotFindCall(pallet_index, call_index)),
 		};
 
 		// Decode each of the argument values in the extrinsic:
-		let mut arguments = vec![];
-		for arg in call.args() {
-			let ty = self.metadata.types().resolve(arg.id()).ok_or_else(|| DecodeError::CannotFindType(arg.id()))?;
-			let val = match decode_type(data, ty, self.metadata.types()) {
-				Ok(val) => val,
-				Err(err) => return Err(err.into()),
-			};
-			arguments.push(val);
-		}
+		let arguments: Vec<_> = variant
+			.fields()
+			.iter()
+			.map(|field| {
+				let type_id = field.ty().id();
+				let ty = self.metadata.types().resolve(type_id).ok_or(DecodeError::CannotFindType(type_id))?;
+				decode_type(data, ty, self.metadata.types()).map_err(DecodeError::DecodeTypeError)
+			})
+			.collect::<Result<_, _>>()?;
+
+		Ok(CallData { pallet_name, ty: variant, arguments })
+	}
 
-		let ext = Extrinsic { pallet: pallet_name.to_owned(), call: call.name().to_owned(), signature, arguments };
+	/// Decode the SCALE encoded data that you would get signed, which conceptually takes the form
```

Comment:
The "get signed" part confuses me a bit, do you mean this decodes a signed extrinsic or…? Maybe I'm just not sure what a `SignerPayload` is.

Reply:
SignerPayload is what you sign to get the signature that you then use to construct the signed extrinsic.